nx/test/nx/optional_test.exs

```diff
@@ -72,23 +86,86 @@ defmodule Nx.OptionalTest do
     end
   end
 
-  describe "optional callbacks" do
+  defmodule DefnInspect do
+    import Nx.Defn, only: [defn: 2]
+
+    defn det(t) do
+      t
+      |> Nx.LinAlg.determinant()
+      |> Nx.sum()
+      |> inspect_expr()
+    end
+  end
+
+  describe "optional callbacks (def)" do
     test "calls the default impl if specific is not present" do
-      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: NonSolveTestBackend)
-      b = Nx.tensor([1, 2, 3], backend: NonSolveTestBackend)
+      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: NonCustomImplTestBackend)
+      b = Nx.tensor([1, 2, 3], backend: NonCustomImplTestBackend)
 
       Nx.LinAlg.solve(a, b)
 
       assert_receive :called_default_impl
     end
 
     test "calls the custom impl if it is present" do
-      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: SolveTestBackend)
-      b = Nx.tensor([1, 2, 3], backend: SolveTestBackend)
+      a = Nx.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]], backend: CustomImplTestBackend)
+      b = Nx.tensor([1, 2, 3], backend: CustomImplTestBackend)
 
       Nx.LinAlg.solve(a, b)
 
       assert_receive :called_custom_impl
     end
   end
+
+  describe "optional callbacks (defn)" do
+    test "calls the custom impl if it exists" do
+      a = Nx.tensor([[1, 1], [2, 1]], backend: CustomImplTestBackend)
+
+      Nx.LinAlg.determinant(a)
+
+      assert_receive :called_custom_impl
+    end
+
+    test "calls the default impl if custom does not exist" do
+      a = Nx.tensor([[1, 1], [2, 1]], backend: NonCustomImplTestBackend)
+
+      Nx.LinAlg.determinant(a)
+
+      assert_receive {:called_default_impl, :reshape}
+    end
+  end
+
+  describe "inspect" do
+    test "works with defn call" do
+      assert ExUnit.CaptureIO.capture_io(fn ->
+               DefnInspect.det(Nx.iota({3, 3}))
+             end) == """
+             #Nx.Tensor<
+               f32
+             \s\s
+               Nx.Defn.Expr
+               parameter a:0                            s64[3][3]
+               b = determinant a                        f32
+               c = sum b, axes: nil, keep_axes: false   f32
+             >
+             """
+    end
+
+    test "works with direct call" do
```

Comment:
@josevalim I'm not sure if we need to support this use case, but this call yields the unrolled default implementation instead of the expected expression. I'm still investigating why in the meantime